### PR TITLE
Configure linux bridge vlan-filtering manually

### PR DESCRIPTION
**Pull request by qinqon**
Only future version of nmstate/[NM](https://github.com/nmstate/nmstate/pull/440) supports linux bridges vlan filtering,
to bypass this problem this PR calls iproute commands ip/bridge to set the
vlan_filtering flag and add all vlans to the bridges itself and to outbound
interfaces.

### CODE REVIEW

**Comment by dankenigsberg** [comment]
> Only newere version of nmstate/NM supports linux bridges vlan filtering,
> to bypass this problem this PR calls iproute commands ip/bridge to set the
> vlan_filtering flag and add all vlans to the bridges itself and to outbound
> interfaces.
> 
> Signed-off-by: Quique Llorente [ellorent@redhat.com](mailto:ellorent@redhat.com)

newere -> future,  and add a link to Miguel's PR.

File: deploy/operator.yaml [hunk 1]
```diff
@@ -14,6 +14,7 @@ spec:
         app: kubernetes-nmstate
         name: nmstate-handler
     spec:
+      hostNetwork: true
```
Comment (dankenigsberg) [inline deploy/operator.yaml]: please add a comment that we add this only for this hack.

File: build/bin/vlan-filtering [hunk 2]
```diff
@@ -0,0 +1,23 @@
+#!/bin/bash -e
+IFS=$'\n'
+sed_expression="^.*: (.*) state .* : .*master (.*) state.*$"
+vlan_range=1-4094
+for port in $(bridge link show | grep -v @); do
+   outbound_name=$(echo $port | sed -r "s/$sed_expression/\1/")
+   bridge_name=$(echo $port | sed -r "s/$sed_expression/\2/")
+   bridges=$(printf $bridge_name\\n$bridges | uniq)
```
Comment (dankenigsberg) [inline build/bin/vlan-filtering]: a safer way to list all host bridges:
```
ip -o link show type bridge
```

File: build/bin/vlan-filtering [hunk 3]
```diff
@@ -0,0 +1,23 @@
+#!/bin/bash -e
+IFS=$'\n'
+sed_expression="^.*: (.*) state .* : .*master (.*) state.*$"
+vlan_range=1-4094
+for port in $(bridge link show | grep -v @); do
+   outbound_name=$(echo $port | sed -r "s/$sed_expression/\1/")
```
Comment (dankenigsberg) [inline build/bin/vlan-filtering]: I think you are going to capture here also `tap` devices, which you should not.

**Comment by qinqon** [comment]
Bonding e2e test is not working I am on it, but we can start review.

File: pkg/helper/client.go [hunk 4]
```diff
@@ -102,5 +113,13 @@ func ApplyDesiredState(nodeNetworkState *nmstatev1alpha1.NodeNetworkState) (stri
 	if len(desiredState) == 0 {
 		return "Ignoring empty desired state", nil
 	}
-	return set(string(nodeNetworkState.Spec.DesiredState))
+	outputSet, err := set(string(nodeNetworkState.Spec.DesiredState))
```
Comment (dankenigsberg) [inline pkg/helper/client.go]: you might be better off parsing the desired state yaml, iterating over desired bridges and the interfaces explicitly set inside them. please consider.

File: build/bin/vlan-filtering (hunk 2 above)
Comment (qinqon) [reply build/bin/vlan-filtering]: I am setting vlan filtering config to bridges that have a port to outbound interface filtering ports with "grep -v @".
ip -o link show type bridge give you all bridges even if they don't have a port to an outbound interface.

File: pkg/helper/client.go (hunk 4 above)
Comment (qinqon) [reply pkg/helper/client.go]: Then maybe we want to reconsider this https://github.com/nmstate/kubernetes-nmstate/pull/140, when NM/nmstate is ready customers will configure vlan-filtering explicitly at k8s-nmstate, with the above PR we can provide that before having nmstate/NM

File: build/bin/vlan-filtering (hunk 3 above)
Comment (qinqon) [reply build/bin/vlan-filtering]: Isn't VMIs tap interfaces at different network namespaces ? or you mean VM out from kubevirt ?

File: pkg/helper/client.go (hunk 4 above)
Comment (qinqon) [reply pkg/helper/client.go]: In case we want to parse desiredState we all need to ignore bridges with "state: absent", would be nice to be able to ask nmstatectl about interfaces managed by it.

Comment (qinqon) [reply pkg/helper/client.go]: After talking with nmstate as expected we are the owners so we should know what interface do we manage, so parsing desiredState have to be, since #140 maybe is too much work.

File: build/bin/vlan-filtering (hunk 3 above)
Comment (dankenigsberg) [reply build/bin/vlan-filtering]: you are right about `tap`, but for each VMI you'd see here a half of a `veth` pair which you should not touch.

File: build/bin/vlan-filtering (hunk 2 above)
Comment (dankenigsberg) [reply build/bin/vlan-filtering]: I think that you should set `vlan_filtering=1` even on floating nic-less bridges.

File: pkg/helper/bridges.go [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+
+func getBridgesUp(desiredState nmstatev1alpha1.State) ([]string, error) {
+	foundBridges := []string{}
+
+	desiredStateYaml, err := yaml.YAMLToJSON([]byte(desiredState))
+	if err != nil {
+		return foundBridges, fmt.Errorf("error converting desiredState to JSON: %v", err)
+	}
+
+	queryResults := gjson.ParseBytes(desiredStateYaml).
```
Comment (phoracek) [inline pkg/helper/bridges.go]: Interesting approach :D I'd expect us to convert it to native Go structure and then iterate that. Won't pushback since it is a temporary workaround.

File: pkg/helper/bridges_test.go [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+      options:
+        miimon: '120'
+`)
+	noBridgesUp = nmstatev1alpha1.State(`interfaces:
+  - name: eth1
+    type: ethernet
+    state: up
+  - name: br1
+    type: linux-bridge
+    state: down
```
Comment (phoracek) [inline pkg/helper/bridges_test.go]: Is state mandatory? If not, what is the default?

File: pkg/helper/bridges_test.go [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+	})
+	Context("when desired state is not a yaml", func() {
+		BeforeEach(func() {
+			desiredState = badYaml
+		})
+		It("should return error", func() {
+			Expect(err).To(HaveOccurred())
+		})
+	})
+	Context("when desired state is empty", func() {
```
Comment (phoracek) [inline pkg/helper/bridges_test.go]: could you handle all non-failing ones via a table?

File: test/e2e/nodenetworkstate_test.go [hunk 8]
```diff
@@ -96,7 +96,6 @@ var _ = Describe("NodeNetworkState", func() {
 
 				// Let's clean the bridge directly in the node
 				// bypassing nmstate
-				deleteConnectionAtNodes("eth1")
```
Comment (phoracek) [inline test/e2e/nodenetworkstate_test.go]: how is this related?

File: pkg/helper/client.go [hunk 9]
```diff
@@ -102,5 +113,17 @@ func ApplyDesiredState(nodeNetworkState *nmstatev1alpha1.NodeNetworkState) (stri
...
```
Comment (phoracek) [inline pkg/helper/client.go]: setOutput

Comment (phoracek) [inline pkg/helper/client.go]: please add an empty line above

File: pkg/helper/client.go [hunk 10]
```diff
@@ -102,5 +113,17 @@ func ApplyDesiredState(nodeNetworkState *nmstatev1alpha1.NodeNetworkState) (stri
...
+	if err != nil {
+		return outputSet, err
+	}
+	bridgesUp, err := getBridgesUp(nodeNetworkState.Spec.DesiredState)
```
Comment (phoracek) [inline pkg/helper/client.go]: empty line above (after error check), the same for the rest of the function

Comment (phoracek) [inline pkg/helper/client.go]: please add a comment referencing the issue

File: pkg/helper/client.go [hunk 11]
```diff
@@ -102,5 +113,17 @@ func ApplyDesiredState(nodeNetworkState *nmstatev1alpha1.NodeNetworkState) (stri
...
+	if err != nil {
+		return "", fmt.Errorf("error retrieving up bridges from desired state: %v", err)
```
Comment (phoracek) [inline pkg/helper/client.go]: This and the next return are questionable. We have no way to rollback if our hack after `nmstatectl set` fails. How would we report it in conditions? Which state are we in? I think we should at least log a big warning.

File: pkg/helper/bridges.go (hunk 5 above)
Comment (qinqon) [reply pkg/helper/bridges.go]: I like this approach better for this cases, even for future policies implementation when you have to query the nmstate yaml to get values.

File: pkg/helper/bridges_test.go (hunk 6 above)
Comment (qinqon) [reply pkg/helper/bridges_test.go]: I think it's mandatory https://github.com/nmstate/nmstate/blob/master/libnmstate/schemas/operational-state.yaml#L71 it does not have minimun: 0

File: test/e2e/nodenetworkstate_test.go (hunk 8 above)
Comment (qinqon) [reply test/e2e/nodenetworkstate_test.go]: Now eth1 is a real secondary nic from kubevirtci, to emulate a real outbound interface.

Comment (phoracek) [reply test/e2e/nodenetworkstate_test.go]: We still create a connection profile for it, therefore we should clean up after the test is done. I will not remove the interface.

File: pkg/helper/bridges.go (hunk 5 above)
Comment (phoracek) [reply pkg/helper/bridges.go]: Granted we keep it as bytes, you might be right. Would make it way easier to implement for sure.

File: test/e2e/nodenetworkstate_test.go (hunk 8 above)
Comment (qinqon) [reply test/e2e/nodenetworkstate_test.go]: We cannot remove it

Comment (phoracek) [reply test/e2e/nodenetworkstate_test.go]: Why not? What happens if you do? Who handles clean-up after the test?

File: pkg/helper/client.go (hunk 11 above)
Comment (qinqon) [reply pkg/helper/client.go]: We need to run this after nmstate since, we create the bridges there, the proper option would be to use commit/rollback and do the rollback in case vlan-filtering fails, so we consider the two operatoins as one.

For now I am going to add a big warning

Is that ok ?

File: pkg/helper/bridges_test.go (hunk 7 above)
Comment (qinqon) [reply pkg/helper/bridges_test.go]: This is temporary workaround don't want to spend more time on it, prefer to use time on e2e

File: build/bin/vlan-filtering (hunk 2 above)
Comment (qinqon) [reply build/bin/vlan-filtering]: I am setting it now at all bridges from desiredState so I cover that, I am going to mark this as resolved.

File: pkg/helper/client.go (hunk 4 above)
Comment (qinqon) [reply pkg/helper/client.go]: I have add filtering by desiredState bridges so I am resolving this.

File: pkg/helper/bridges_test.go (hunk 7 above)
Comment (phoracek) [reply pkg/helper/bridges_test.go]: "temporary" is the biggest lie! But ok, it's not that bad. But next time, please consider `DescribeTable`

File: pkg/helper/client.go (hunk 11 above)
Comment (phoracek) [reply pkg/helper/client.go]: Yes, big warning is fine. And a TODO to handle the rollback once we have needed tools in `nmstatectl`.

File: test/e2e/nodenetworkstate_test.go (hunk 8 above)
Comment (qinqon) [reply test/e2e/nodenetworkstate_test.go]: Humm we should remove them, don't know why I delete the delete, adding them back.

File: pkg/helper/client.go (hunk 11 above)
Comment (qinqon) [reply pkg/helper/client.go]: Well as commented it's going to re-concile if vlan-filtering fails so that's like a big warning, so added the TODO about commit/rollback

File: build/bin/vlan-filtering (hunk 3 above)
Comment (qinqon) [reply build/bin/vlan-filtering]: But I am filtering out veths since they appear with "@" so we are not seting vlans for them VMIs or normal pods.

File: build/bin/vlan-filtering [hunk 12]
```diff
@@ -0,0 +1,16 @@
+#!/bin/bash -e
+
+vlan_range=1-4094
+
+for bridge in $@; do
+   ip link set dev $bridge type bridge vlan_filtering 1
+   bridge vlan add dev $bridge vid $vlan_range self
+
+   IFS=$'\n'
+   for port in $(bridge link show $bridge | grep -v @ | grep "state forwarding"); do
```
Comment (phoracek) [inline build/bin/vlan-filtering]: not sure if it helps and if it is available on RHEL, but you can use -json to get structured output

Comment (qinqon) [reply build/bin/vlan-filtering]: Thanks! it's working bridge -j vlan show, that makes e2e easier !!!

File: build/bin/vlan-filtering [hunk 13]
```diff
@@ -0,0 +1,16 @@
+#!/bin/bash -e
+
+vlan_range=1-4094
```
Comment (maiqueb) [inline build/bin/vlan-filtering]: There's one thing that puzzles me with the selected range.

So you activate vlan filtering **and** whitelist **all** the vlans (on all ports) ? It's unclear to me what are the intentions.

One possibility is that there are more "users" of that bridge and you don't want to impair their networking, but still, would like to point this out.

Comment (dankenigsberg) [reply build/bin/vlan-filtering]: We need to allow all incoming traffic from the host interface. When we configure the bridge we have no idea which NetAttDef are going to be defined in the cluster (nor do we want to know)

**Comment by qinqon** [comment]
@dankenigsberg @phoracek @maiqueb I have manually confirmed that nmstate does not remove the ports not configured by it (AKA veths) so we are good there.

File: build/bin/vlan-filtering (hunk 3 above)
Comment (qinqon) [reply build/bin/vlan-filtering]: @dankenigsberg so we are good ? I don't see any change needed since we are not touching veths at all here

**Comment by maiqueb** [comment]
> Remember that by default it also sets vlan 1 as pvid, meaning you will get a 'freely' provisioned vlan with the following config:
> "1 PVID Egress Untagged"
> 
> Not sure you want that, or explicitly want to get rid of it.

Since you're redefining the vlan 1, the default is overwritten, disregard this comment.
